fl4health/clients/fenda_client.py

```diff
@@ -1,12 +1,226 @@
+            self.old_local_module = self.clone_and_freeze_model(self.model.local_module)
+            self.old_global_module = self.clone_and_freeze_model(self.model.global_module)
+
+        return super().get_parameters(config)
+
+    def set_parameters(self, parameters: NDArrays, config: Config) -> None:
+
+        # Set the parameters of the model
+        super().set_parameters(parameters, config)
+
+        # Save the parameters of the aggregated global model
+        assert isinstance(self.model, FendaModel)
+        if self.perfcl_loss_weights:
+            self.aggregated_global_module = self.clone_and_freeze_model(self.model.global_module)
+
+    def get_cosine_similarity_loss(self, local_features: torch.Tensor, global_features: torch.Tensor) -> torch.Tensor:
+        """
+        Cosine similarity loss aims to minimize the similarity among current local features and current global
+        features of fenda model.
+        """
+        assert len(local_features) == len(global_features)
+        return torch.abs(self.cos_sim(local_features, global_features)).mean()
+
+    def compute_contrastive_loss(
+        self, features: torch.Tensor, positive_pairs: torch.Tensor, negative_pairs: torch.Tensor
+    ) -> torch.Tensor:
+        """
+        Contrastive loss aims to enhance the similarity between the features and their positive pairs
+        while reducing the similarity between the features and their negative pairs.
+        """
+        assert self.temperature is not None
+        assert len(features) == len(positive_pairs)
+        posi = self.cos_sim(features, positive_pairs)
+        logits = posi.reshape(-1, 1)
+        assert len(features) == len(negative_pairs)
+        nega = self.cos_sim(features, negative_pairs)
+        logits = torch.cat((logits, nega.reshape(-1, 1)), dim=1)
+        logits /= self.temperature
+        labels = torch.zeros(features.size(0)).to(self.device).long()
+
+        return self.ce_criterion(logits, labels)
+
+    def get_contrastive_loss(
+        self, local_features: torch.Tensor, old_local_features: torch.Tensor, global_features: torch.Tensor
+    ) -> torch.Tensor:
+        """
+        Compute contrastive over current local features (z_p) with old local features (hat{z_p}) as positive pairs
+        and current global features (z_s) as negative pairs.
+        """
+        return self.compute_contrastive_loss(
+            features=local_features,  # (z_p)
+            positive_pairs=old_local_features,  # (\hat{z_p})
+            negative_pairs=global_features,  # (z_s)
+        )
+
+    def get_perfcl_loss(
+        self,
+        local_features: torch.Tensor,
+        old_local_features: torch.Tensor,
+        global_features: torch.Tensor,
+        old_global_features: torch.Tensor,
+        aggregated_global_features: torch.Tensor,
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        """
+        Perfcl loss consists of two contrastive loss computation.
```

Comment:
Do you mind putting a link to the arXiv paper in this comment to draw a line from the variable notation to the paper itself?